rasa/engine/validation.py

```diff
@@ -0,0 +1,324 @@
+            f"reference by removing the quotes around the type "
+            f"(e.g. 'def foo() -> \"int\"' becomes 'def foo() -> int'. and make sure "
+            f"all type annotations can be resolved during runtime."
+        )
+
+
+def _get_fn(node_name: Text, uses: Type[GraphComponent], method_name: Text) -> Callable:
+    fn = getattr(uses, method_name, None)
+    if fn is None:
+        raise GraphSchemaValidationException(
+            f"Node '{node_name}' uses graph component '{uses.__name__}' which does not "
+            f"have the specified "
+            f"method '{method_name}'. Please make sure you're either using "
+            f"the right graph component or specifying a valid method "
+            f"for this component."
+        )
+    return fn
+
+
+def _validate_run_fn(
+    node_name: Text,
+    node: SchemaNode,
+    run_fn_params: Dict[Text, ParameterInfo],
+    run_fn_return_type: Any,
+    is_train_graph: bool,
+) -> None:
+    _validate_types_of_reserved_keywords(run_fn_params, node_name, node, node.fn)
+    _validate_run_fn_return_type(node_name, node, run_fn_return_type, is_train_graph)
+
+    for param_name in _required_args(run_fn_params):
+        if param_name not in node.needs:
+            raise GraphSchemaValidationException(
+                f"Node '{node_name}' uses a component '{node.uses.__name__}' which "
+                f"needs the param '{param_name}' to be provided to its method "
+                f"'{node.fn}'. Please make sure to specify the parameter in "
+                f"the node's 'needs' section."
+            )
+
+
+def _required_args(fn_params: Dict[Text, ParameterInfo]) -> Set[Text]:
+    keywords = set(_keywords_expected_types)
+    return {
+        param_name
+        for param_name, param in fn_params.items()
+        if not param.has_default and not param.is_kwargs and param_name not in keywords
+    }
+
+
+def _validate_run_fn_return_type(
+    node_name: Text, node: SchemaNode, return_type: Type, is_training: bool
+) -> None:
+    if return_type == inspect.Parameter.empty:
+        raise GraphSchemaValidationException(
+            f"Node '{node_name}' uses a component '{node.uses.__name__}' whose "
+            f"method '{node.fn}' does not have a type annotation for "
+            f"its return value. Type annotations are required for all graph "
+            f"components to validate the graph's structure."
+        )
+
+    if is_training and not isinstance(return_type, Fingerprintable):
+        raise GraphSchemaValidationException(
+            f"Node '{node_name}' uses a component '{node.uses.__name__}' whose method "
+            f"'{node.fn}' does not return a fingerprintable "
+            f"output. This is required for caching. Please make sure you're "
+            f"using a return type which implements the "
+            f"'{Fingerprintable.__name__}' protocol."
+        )
+
+
+def _validate_types_of_reserved_keywords(
```

Comment:
yep, but that wouldn't work with custom components. I didn't find a way to cleanly call mypy from Python code 😬

Reply:
> yep, but that wouldn't work with custom components.

Why not? It would enforce that they implement the methods of the protocol with correct parameters and types.

> I didn't find a way to cleanly call mypy from Python code 😬

Yeah I also looked when first looking at protocols and couldn't find anything nice.

Reply:
> Why not? It would enforce that they implement the methods of the protocol with correct parameters and types.

True, but it would only work for `create` and `load` as the `run` fn is dynamic

Reply:
Yup true - but we never have these keywords in `run` right?
Anyway pointless debate - this is good 👍